### PR TITLE
Local object usage

### LINKED ISSUE #31: Should not lint on not locally used `@export`ed functions
## box.linters version

0.0.0.9006

## Source code to lint

```R
#' @export
public_function <- function() {
  private_function()
}

private_function <- function() {
  \"private function\"
}
```

## Lint command

```R
lintr::lint(code, linters = lintr::linters_with_defaults(defaults = box.linters::rhino_default_linters))
```

## Lint result

```
<text>:3:1: warning: [unused_declared_func_linter] Declared function unused.
public_function <- function() {
^~~~~~~~~~~~~~~
```

## Expected result

Should not lint on `@export`ed functions.

### LINKED ISSUE #32: Should not lint on locally unused @exported data objects
## box.linters version

0.0.0.9006

## Source code to lint

```R
#' @export
public_data <- c(1, 3)
```

## Lint command

```R
lintr::lint(code, linters = lintr::linters_with_defaults(defaults = box.linters::rhino_default_linters))
```

## Expected result

Should not lint on unused `public_data`.

### LINKED ISSUE #33: Should lint on not-locally used functions
## box.linters version

0.0.0.9006

## Source code to lint

```R
#' @export
public_function <- function() {
  \"exported function\"
}

private_function <- function() {

}
```

## Lint command

```R
lintr::lint(code, linters = lintr::linters_with_defaults(defaults = box.linters::rhino_default_linters))
```

## Lint result
```
<text>:7:1: warning: [unused_declared_func_linter] Declared function unused.
private_function <- function() {
^~~~~~~~~~~~~~~~
```

## Expected result

Should lint on not locally used functions

### LINKED ISSUE #34: Should lint on not-locally used private data
## box.linters version

0.0.0.9006

## Source code to lint

```R
#' @export
public_function <- function() {
  \"exported function\"
}

private_data <- c(1, 3)
```

## Lint command

```R
lintr::lint(code, linters = lintr::linters_with_defaults(defaults = box.linters::rhino_default_linters))
```

## Expected result

Should lint on unused `private_data`.

### PR DESCRIPTION
**Pull request by radbasa**
Extends from PR #39 

- closes #31 
- closes #32 
- closes #33 
- closes #34

## Changes
- allows `@export`ed functions and data objects to be unused locally
- prevents unused private functions and data objects

## Definition of Done
- [ ] The change is thoroughly documented.
- [ ] The CI passes (`R CMD check`, linter, unit tests, spelling).
- [ ] Any generated files have been updated (e.g. `.Rd` files with `roxygen2::roxygenise()`)

### CODE REVIEW

**Comment by jakubnowicki** [comment]
Linter throws style errors on `shiny` objects like `input` and `output`:
```
app/view/mod_app_table.R:44:12: warning: [unused_declared_object_linter] Declared function/object unused.
    output$app_table <- renderReactable({
```

**Comment by jakubnowicki** [comment]
Linter throws errors on parameters of `server` function (so ones that can be used in another module):
```r
#' @export
server <- function(id, state) {
  moduleServer(id, function(input, output, session) {

    state$selected_job <- reactive({
```
```
app/view/mod_job_list.R:86:11: warning: [unused_declared_object_linter] Declared function/object unused.
    state$selected_job <- reactive({
```

File: tests/testthat/shiny/simple/app.R [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+  output$distPlot <- renderPlot({
+    x    <- faithful[, 2]
+    bins <- seq(min(x), max(x), length.out = input$bins + 1)
+
+    hist(x, breaks = bins, col = "darkgray", border = "white",
+         xlab = "Waiting time to next eruption (in mins)",
+         main = "Histogram of waiting times")
+  })
+}
```
Comment (radbasa) [inline tests/testthat/shiny/simple/app.R:34]: Our `unused_declared_object_linter()` sees the following assigned objects:

```
$xml_nodes
{xml_nodeset (4)}
[1] <SYMBOL line1="3" col1="1" line2="3" col2="6" start="190" end="195">server</SYM ...
[2] <SYMBOL line1="5" col1="10" line2="5" col2="17" start="325" end="332">distPlot< ...
[3] <SYMBOL line1="6" col1="5" line2="6" col2="5" start="383" end="383">x</SYMBOL>
[4] <SYMBOL line1="7" col1="5" line2="7" col2="8" start="446" end="449">bins</SYMBOL>

$text
[1] "server"   "distPlot" "x"        "bins"    
```

And so, linter throws the following:

```
<text>:5:10: warning: [box.linters::unused_declared_object_linter] Declared function/object unused.
  output$distPlot <- renderPlot({
```

The solution at this moment is to ignore assignments to `$element`. In this case, ignore `output$distPlot` as an object that should be checked for local use. The consequence is this affects all other list assignment code.

File: tests/testthat/shiny/simple/app.R [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                  value = 30)
+    ),
+
+    mainPanel(
+      plotOutput("distPlot")
+    )
+  )
+)
+
+server <- function(input, output) {
```
Comment (jakubnowicki) [inline tests/testthat/shiny/simple/app.R:24]: Please make sure that the complex e2e test will include more cases (like the ones from my [comment](https://github.com/Appsilon/box.linters/pull/40#issuecomment-2110503186)).

**Review by jakubnowicki — APPROVED** [review]
Works as expected, good job.

**Pull request merged**